File: tests/src/gui/testqgsfiledownloader.cpp
/***************************************************************************
    testqgsfilefiledownloader.cpp
     --------------------------------------
    Date                 : 11.8.2016
    Copyright            : (C) 2016 Alessandro Pasotti
    Email                : apasotti at boundlessgeo dot com
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/


#include "qgstest.h"
#include <QObject>
#include <QTemporaryFile>
#include <QTemporaryDir>
#include <QUrl>
#include <QEventLoop>
#include <QTimer>

#include <qgsapplication.h>
#include <qgsfiledownloader.h>

class TestQgsFileDownloader: public QObject
{
    Q_OBJECT
  public:
    TestQgsFileDownloader() = default;

  public slots:
    //! Called when the download has completed successfully
    void downloadCompleted()
    {
      mCompleted = true;
    }
    //! Called when the download exits
    void downloadExited()
    {
      mExited = true;
    }
    //! Called when the download was canceled by the user
    void downloadCanceled()
    {
      mCanceled = true;
    }
    //! Called when an error makes the download fail
    void downloadError( QStringList errorMessages )
    {
      mError = true;
      errorMessages.sort();
      mErrorMessage = errorMessages.join( QLatin1Char( ';' ) );
    }
    //! Called when data ready to be processed
    void downloadProgress( qint64 bytesReceived, qint64 bytesTotal )
    {
      Q_UNUSED( bytesReceived );
      Q_UNUSED( bytesTotal );
      mProgress = true;
    }

  private slots:
    void initTestCase(); // will be called before the first testfunction is executed.
    void cleanupTestCase(); // will be called after the last testfunction was executed.
    void init(); // will be called before each testfunction is executed.
    void cleanup(); // will be called after every testfunction.

    void testValidDownload();
    void testInValidDownload();
    void testInvalidFile();
    void testCanceledDownload();
    void testInvalidUrl();
    void testBlankUrl();
    void testLacksWritePermissionsError();
#ifndef QT_NO_SSL
    void testSslError_data();
    void testSslError();
#endif

  private:
    void makeCall( QUrl url, QString fileName, bool cancel = false );
    QTemporaryFile *mTempFile = nullptr;
    QString mErrorMessage;
    bool mCanceled =  false ;
    bool mProgress =  false ;
    bool mError =  false ;
    bool mCompleted =  false ;
    bool mExited =  false ;
    QgsFileDownloader *mFileDownloader = nullptr;
};

void TestQgsFileDownloader::makeCall( QUrl url, QString fileName, bool cancel )
{
  QEventLoop loop;

  mFileDownloader = new QgsFileDownloader( url, fileName );
  connect( mFileDownloader, &QgsFileDownloader::downloadCompleted, this, &TestQgsFileDownloader::downloadCompleted );
  connect( mFileDownloader, &QgsFileDownloader::downloadCanceled, this, &TestQgsFileDownloader::downloadCanceled );
  connect( mFileDownloader, &QgsFileDownloader::downloadExited, this, &TestQgsFileDownloader::downloadExited );
  connect( mFileDownloader, &QgsFileDownloader::downloadError, this, &TestQgsFileDownloader::downloadError );
  connect( mFileDownloader, &QgsFileDownloader::downloadProgress, this, &TestQgsFileDownloader::downloadProgress );

  connect( mFileDownloader, &QgsFileDownloader::downloadExited, &loop, &QEventLoop::quit );

  if ( cancel )
    QTimer::singleShot( 1000, mFileDownloader, &QgsFileDownloader::cancelDownload );

  loop.exec();

}

void TestQgsFileDownloader::initTestCase()
{
  QgsApplication::init();
  QgsApplication::initQgis();

}

void TestQgsFileDownloader::cleanupTestCase()
{
  QgsApplication::exitQgis();
}

void TestQgsFileDownloader::init()
{
  mErrorMessage.clear();
  mCanceled = false;
  mProgress = false;
  mError = false;
  mCompleted = false;
  mExited = false;
  mTempFile = new QTemporaryFile();
  QVERIFY( mTempFile->open() );
  mTempFile->close();
}



void TestQgsFileDownloader::cleanup()
{
  delete mTempFile;
}

void TestQgsFileDownloader::testValidDownload()
{
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( QStringLiteral( "http://www.qgis.org" ) ), mTempFile->fileName() );
  QVERIFY( mExited );
  QVERIFY( mCompleted );
  QVERIFY( mProgress );
  QVERIFY( !mError );
  QVERIFY( !mCanceled );
  QVERIFY( mTempFile->size() > 0 );
}

void TestQgsFileDownloader::testInValidDownload()
{
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( QStringLiteral( "http://www.doesnotexistofthatimsure.qgis" ) ), mTempFile->fileName() );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
  QVERIFY( mTempFile->size() == 0 );
  QCOMPARE( mErrorMessage, QString( "Download failed: Host www.doesnotexistofthatimsure.qgis not found" ) );
}

void TestQgsFileDownloader::testCanceledDownload()
{
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( QStringLiteral( "https://github.com/qgis/QGIS/archive/master.zip" ) ), mTempFile->fileName(), true );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( !mError );
  QVERIFY( mProgress );
  QVERIFY( mCanceled );
  QVERIFY( !mTempFile->exists() );
}

void TestQgsFileDownloader::testInvalidFile()
{
  makeCall( QUrl( QStringLiteral( "https://github.com/qgis/QGIS/archive/master.zip" ) ), QString() );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
  QCOMPARE( mErrorMessage, QString( "No output filename specified" ) );
}

void TestQgsFileDownloader::testInvalidUrl()
{
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( QStringLiteral( "xyz://www" ) ), mTempFile->fileName() );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
  QCOMPARE( mErrorMessage, QString( "Download failed: Protocol \"xyz\" is unknown" ) );
}

void TestQgsFileDownloader::testBlankUrl()
{
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( QString() ), mTempFile->fileName() );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
  QCOMPARE( mErrorMessage, QString( "Download failed: Protocol \"\" is unknown" ) );
}

#ifndef QT_NO_SSL
void TestQgsFileDownloader::testSslError_data()
{
  QTest::addColumn<QString>( "url" );
  QTest::addColumn<QString>( "result" );

  QTest::newRow( "expired" ) << "https://expired.badssl.com/"
                             << "SSL Errors: ;The certificate has expired";
  QTest::newRow( "self-signed" ) << "https://self-signed.badssl.com/"
                                 << "SSL Errors: ;The certificate is self-signed, and untrusted";
  QTest::newRow( "untrusted-root" ) << "https://untrusted-root.badssl.com/"
                                    << "No certificates could be verified;SSL Errors: ;The issuer certificate of a locally looked up certificate could not be found";
}

void TestQgsFileDownloader::testSslError()
{
  QFETCH( QString, url );
  QFETCH( QString, result );
  QVERIFY( ! mTempFile->fileName().isEmpty() );
  makeCall( QUrl( url ), mTempFile->fileName() );
  QCOMPARE( mErrorMessage, result );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
}

void TestQgsFileDownloader::testLacksWritePermissionsError()
{
  const QTemporaryDir dir;
  QFile tmpDir( dir.path( ) );
  tmpDir.setPermissions( tmpDir.permissions() & ~( QFile::Permission::WriteGroup |  QFile::Permission::WriteUser | QFile::Permission::WriteOther | QFile::Permission::WriteOwner ) );
  QVERIFY( ! tmpDir.isWritable() );
  const QString fileName( dir.path() + '/' + QStringLiteral( "tmp.bin" ) );
  makeCall( QUrl( QStringLiteral( "http://www.qgis.org" ) ), fileName );
  QVERIFY( mExited );
  QVERIFY( !mCompleted );
  QVERIFY( mError );
  QVERIFY( !mCanceled );
  QVERIFY( ! QFileInfo::exists( fileName ) );
}


#endif


QGSTEST_MAIN( TestQgsFileDownloader )
#include "testqgsfiledownloader.moc"

File: tests/src/python/test_qgsfiledownloader.py
"""
Test the QgsFileDownloader class

Run test with:
LC_ALL=en_US.UTF-8 ctest -V -R PyQgsFileDownloader

.. note:: This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
"""

import os
import tempfile
import unittest
from functools import partial

from qgis.core import QgsFileDownloader
from qgis.PyQt.QtCore import QEventLoop, QUrl
from qgis.testing import QgisTestCase, start_app

__author__ = "Alessandro Pasotti"
__date__ = "08/11/2016"
__copyright__ = "Copyright 2016, The QGIS Project"

start_app()


class TestQgsFileDownloader(QgisTestCase):
    """
    This class tests the QgsFileDownloader class
    """

    def _make_download(self, url, destination, cancel=False):
        self.completed_was_called = False
        self.error_was_called = False
        self.canceled_was_called = False
        self.progress_was_called = False
        self.exited_was_called = False

        loop = QEventLoop()

        downloader = QgsFileDownloader(QUrl(url), destination)
        downloader.downloadCompleted.connect(partial(self._set_slot, "completed"))
        downloader.downloadExited.connect(partial(self._set_slot, "exited"))
        downloader.downloadCanceled.connect(partial(self._set_slot, "canceled"))
        downloader.downloadError.connect(partial(self._set_slot, "error"))
        downloader.downloadProgress.connect(partial(self._set_slot, "progress"))

        downloader.downloadExited.connect(loop.quit)

        if cancel:
            downloader.downloadProgress.connect(downloader.cancelDownload)

        loop.exec()

    @unittest.skipIf(
        os.environ.get("QGIS_CONTINUOUS_INTEGRATION_RUN", "true"),
        "Test with http://www.qgis.org unstable. Needs local server.",
    )
    def test_validDownload(self):
        """Tests a valid download"""
        destination = tempfile.mktemp()
        self._make_download("http://www.qgis.org", destination)
        self.assertTrue(self.exited_was_called)
        self.assertTrue(self.completed_was_called)
        self.assertTrue(self.progress_was_called)
        self.assertFalse(self.canceled_was_called)
        self.assertFalse(self.error_was_called)
        self.assertTrue(os.path.isfile(destination))
        self.assertGreater(os.path.getsize(destination), 0)

    def test_inValidDownload(self):
        """Tests an invalid download"""
        destination = tempfile.mktemp()
        self._make_download("http://www.doesnotexistofthatimsure.qgis", destination)
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called)
        self.assertTrue(self.progress_was_called)
        self.assertFalse(self.canceled_was_called)
        self.assertTrue(self.error_was_called)
        self.assertEqual(
            self.error_args[1],
            ["Download failed: Host www.doesnotexistofthatimsure.qgis not found"],
        )
        self.assertFalse(os.path.isfile(destination))

    @unittest.skipIf(
        os.environ.get("QGIS_CONTINUOUS_INTEGRATION_RUN", "true"),
        "Test with http://www.github.com unstable. Needs local server.",
    )
    def test_dowloadCanceled(self):
        """Tests user canceled download"""
        destination = tempfile.mktemp()
        self._make_download(
            "https://github.com/qgis/QGIS/archive/master.zip", destination, True
        )
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called)
        self.assertTrue(self.canceled_was_called)
        self.assertFalse(self.error_was_called)
        self.assertFalse(os.path.isfile(destination))

    def test_InvalidUrl(self):
        destination = tempfile.mktemp()
        self._make_download("xyz://www", destination)
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called)
        self.assertFalse(self.canceled_was_called)
        self.assertTrue(self.error_was_called)
        self.assertFalse(os.path.isfile(destination))
        self.assertEqual(
            self.error_args[1], ['Download failed: Protocol "xyz" is unknown']
        )

    @unittest.skipIf(
        os.environ.get("QGIS_CONTINUOUS_INTEGRATION_RUN", "true"),
        "Test with http://www.github.com unstable. Needs local server.",
    )
    def test_InvalidFile(self):
        self._make_download("https://github.com/qgis/QGIS/archive/master.zip", "")
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called)
        self.assertFalse(self.canceled_was_called)
        self.assertTrue(self.error_was_called)
        self.assertEqual(self.error_args[1], ["No output filename specified"])

    def test_BlankUrl(self):
        destination = tempfile.mktemp()
        self._make_download("", destination)
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called)
        self.assertFalse(self.canceled_was_called)
        self.assertTrue(self.error_was_called)
        self.assertFalse(os.path.isfile(destination))
        self.assertEqual(
            self.error_args[1], ['Download failed: Protocol "" is unknown']
        )

    def ssl_compare(self, name, url, error):
        destination = tempfile.mktemp()
        self._make_download(url, destination)
        msg = f"Failed in {name}: {url}\n"
        self.assertTrue(self.exited_was_called)
        self.assertFalse(self.completed_was_called, msg)
        self.assertFalse(self.canceled_was_called, msg)
        self.assertTrue(self.error_was_called, msg)
        self.assertFalse(os.path.isfile(destination), msg)
        result = sorted(self.error_args[1])
        result = ";".join(result)
        self.assertTrue(
            result.startswith(error), msg + f"- Expected: {error}\n- Actual: {result}\n"
        )

    def test_sslExpired(self):

        # badssl.com is really unstable, so prefer setting up a local instance if you want to reproduce
        # like it's done in CI. To do so, see tests folder README.md, section "Local badssl server"

        expired_url = os.environ.get(
            "QGIS_BADSSL_URL_EXPIRED", "https://expired.badssl.com/"
        )
        print(f"expired url: {expired_url}")
        self.ssl_compare(
            "expired",
            expired_url,
            "SSL Errors: ;The certificate has expired",
        )

        selfsigned_url = os.environ.get(
            "QGIS_BADSSL_URL_SELFSIGNED", "https://self-signed.badssl.com/"
        )
        print(f"selfsigned url: {selfsigned_url}")
        self.ssl_compare(
            "self-signed",
            selfsigned_url,
            "SSL Errors: ;The certificate is self-signed, and untrusted",
        )

        untrusted_url = os.environ.get(
            "QGIS_BADSSL_URL_UNTRUSTED", "https://untrusted-root.badssl.com/"
        )
        print(f"untrusted_url: {untrusted_url}")
        self.ssl_compare(
            "untrusted-root",
            untrusted_url,
            "SSL Errors: ;The root certificate of the certificate chain is self-signed, and untrusted",
        )

    def _set_slot(self, *args, **kwargs):
        # print('_set_slot(%s) called' % args[0])
        setattr(self, args[0] + "_was_called", True)
        setattr(self, args[0] + "_args", args)


if __name__ == "__main__":
    unittest.main()
